### ISSUE TITLE
No ability to set dynamic value for kubeReserved cpu

### ISSUE THREAD

**Issue by jukie** [aws/karpenter-provider-aws#6665]
### Description

**What problem are you trying to solve?**
I'd like the ability to set a common kubelet configuration value for maxPods and kubeReserved.memory but would like to retain the dynamic cpu configuration that's used in the [eks bootstrap.sh
](https://github.com/awslabs/amazon-eks-ami/blob/62c3103c9e282b6f67fbd5865512c962f93ae784/templates/al2/runtime/bootstrap.sh#L280-L289). This is because settings kubelet.maxPods alone doesn't lead to any changes in kubelet.kubeReserved as those values will still be generated based on `/etc/eks/eni-max-pods.txt`

If I only set kubeReserved.memory kubelet will fail to start due to an invalid config and if I override the value in bootstrap.sh during userdata execution Karpenter is then unaware of the actual values when making scheduling decisions.

Is there a way to configure a NodePool in a way that would allow for setting a static value for kubeReserved.memory while also allowing for the dynamic CPU configuration from bootstrap.sh?
```
# Calculates the amount of CPU to reserve for kubeReserved in millicores from the total number of vCPUs available on the instance.
# From the total core capacity of this worker node, we calculate the CPU resources to reserve by reserving a percentage
# of the available cores in each range up to the total number of cores available on the instance.
# We are using these CPU ranges from GKE (https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#node_allocatable):
# 6% of the first core
# 1% of the next core (up to 2 cores)
# 0.5% of the next 2 cores (up to 4 cores)
# 0.25% of any cores above 4 cores
# Return:
#   CPU resources to reserve in millicores (m)
get_cpu_millicores_to_reserve() {
  local total_cpu_on_instance=$(($(nproc) * 1000))
  local cpu_ranges=(0 1000 2000 4000 $total_cpu_on_instance)
  local cpu_percentage_reserved_for_ranges=(600 100 50 25)
  cpu_to_reserve="0"
  for i in "${!cpu_percentage_reserved_for_ranges[@]}"; do
    local start_range=${cpu_ranges[$i]}
    local end_range=${cpu_ranges[(($i + 1))]}
    local percentage_to_reserve_for_range=${cpu_percentage_reserved_for_ranges[$i]}
    cpu_to_reserve=$(($cpu_to_reserve + $(get_resource_to_reserve_in_range $total_cpu_on_instance $start_range $end_range $percentage_to_reserve_for_range)))
  done
  echo $cpu_to_reserve
}
```

**How important is this feature to you?**
Very important and would be willing to contribute this feature

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jukie** [comment]
This probably should've been opened under karpenter-core instead so I've created https://github.com/kubernetes-sigs/karpenter/issues/1518 and can close whichever one makes most sense to keep.

**Comment by jukie** [comment]
Reopening as this may be doable within the aws provider. [This function](https://github.com/aws/karpenter-provider-aws/blob/06103f76d268d27e2d8e3a47a557f463c49d943d/pkg/providers/instancetype/types.go#L369-L402) and/or the logic above it could be adjusted to consider each resource. It currently just checks whether kubelet.kubereserved is non-null and requires everything to be set but if it instead checked each resource that could give the desired result.

**Comment by njtran** [comment]
As discussed in the upstream issue, looking to discuss this in just the provider. Can you give more details on why you want to override one part of the kube-reserved resources, but not all? Please correct me if that's not your request.

**Comment by jukie** [comment]
As you mentioned in https://github.com/kubernetes-sigs/karpenter/issues/1518#issuecomment-2284447020 memory scales with pod density and that's the only thing I'd like to change.
To explain the use case more, currently when setting maxPods as a kubelet config this doesn't lead to any change in kubeReserved memory because bootstrap.sh looks at `/etc/eks/eni-max-pods.txt` 
https://github.com/awslabs/amazon-eks-ami/blob/62c3103c9e282b6f67fbd5865512c962f93ae784/templates/al2/runtime/bootstrap.sh#L510 and will update the kubelet config file.

I'd like the ability to override kubeReserved memory so that it aligns with the proper value for maxPods but would still like CPU or ephemeral storage to use the default calculated values.

To fix this I think Karpenter should always provide kubeReserved values as kubelet extra args and use the default values which Karpenter is already calcuating for any empty fields (ephemeral-storage or cpu in this case).

**Comment by stevehipwell** [comment]
I think this is a duplicate of #1803.

**Comment by jukie** [comment]
Slightly different but depending on the solution could solve both. Technically my issue is that I'm its current form kube-reserves config is all or nothing, and I want karpenter to provide computed defaults for any empty fields.

I don't think solving this through userdata or changes in bootstrap.sh is ideal as there should be a common way to configure across cloud providers or even when using custom Ami's without bootstrap.sh. Additionally, having it computed by Karpenter is preferable to handle node scheduling and make sure the perceived allocatable resources matches actual. That's a separate discussion but I think the vm overhead calculation logic could be tweaked as well.

Karpenter already has the logic for calculating kube-reserved so I think it should simply _always_ provide these to kubelet args.

**Comment by stevehipwell** [comment]
@jukie kube reserved is a kubelet setting so you have to integrate it via user data.

**Comment by jukie** [comment]
Correct, and when a user provides kube reserved values these are already handled that way by Karpenter. What I'm suggesting is that when partial fields are provided the empty ones should have defaults injected or take it further to always provide them.
The main point is that in certain scenarios the Karpenter perceived values will differ from the bootstrap.sh generated values or if using a custom ami without bootstrap.sh these values will go unconfigured.

**Comment by johngmyers** [comment]
I think there are two issues here:

1) karpenter-provider-aws and bootstrap.sh should not each independently calculate kubeReserved. A discrepancy in the two formulas can lead to scheduling problems. Karpenter-provider-aws should provide the result of its calculation in userdata for use by bootstrap.sh

2) The EC2NodeClass `spec.kubeReserved` and `spec.systemReserved` fields should have additional options to specify per-max-pod and per-node additions to the values calculated by the default functions.

Our use case: our nodes have DaemonSets which consume CPU and memory. We want to add these fixed per-node amounts to `kubeReserved` to account for this use. We cannot account for these in the DaemonSet pod resource requests because we do not want our distributed databases to be trashed by kubernetes/kubernetes#102382.

**Comment by stevehipwell** [comment]
@johngmyers based on our use case of a fixed max pods per node we now just provide a static kube reserved for all node types. This keeps Karpenter and bootstrap consistent. I still think there should be dynamic support, but as Karpenter doesn't manage the node sizes directly anything based on node properties need the logic mirroring between Karpenter and bootstrap.

For your specific issue, have you considered using a preemptible DaemonSet that runs alongside your DaemonSet and reserves the resources required for it to be updated?

**Comment by johngmyers** [comment]
We don't have a fixed max pods per node.

I hadn't considered a preemptible Daemonset. I'm not quite sure it has quite the right semantics in the edge cases (being preemptable and all) but it might be close enough.

**Comment by stevehipwell** [comment]
@johngmyers you might need to play about with the disruption config but it should do what you want. I'd be less convinced that increasing the kube reserved value would work, given that this only helps in the case where more resources are being used than you've reserved for your pods; and only if the pods can borrow some of the extra kube reserved resources.

**Comment by jmdeal** [comment]
I'm going to close this out in favor of a more general purpose feature request: https://github.com/aws/karpenter-provider-aws/issues/8742
